### PR TITLE
Uptime monitor blogpost

### PR DESCRIPTION
**Pull request by mies**
blogpost related to the uptime monitor sample app

### CODE REVIEW

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+author: Micha "mies" Hernandez van Leuffen
+tags:
+  - HONC
+  - uptimemonitor
+  - Honojs
+---
+
+import { LinkCard } from "@astrojs/starlight/components";
+
+In this tutorial, we'll build a serverless website uptime monitoring application using Cloudflare Workers. Our application will periodically check websites' availability, track response times, and store the results in a serverless database. We'll use the HONC stack: Hono (a lightweight web framework), OpenTelemetry (for observability), and Cloudflare's suite of serverless products.
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
In this tutorial, we'll build a serverless website uptime monitoring application using Cloudflare Workers. Our application will periodically check websites' availability, track response times, and store the results in a serverless database. We'll use the [HONC stack](https://honc.dev/): Hono (a lightweight web framework), OpenTelemetry (for observability), and Cloudflare's suite of serverless products.
```

Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]: not sure about OpenTelemetry? Should it not mention the components of the HONC stack?

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+* Allow configurable check intervals per website
+* Track response times and HTTP status codes
+* Calculate uptime percentages
+* Provide a RESTful API for managing monitored sites
+* Include a simple web interface to view results
+
+## The Technology Stack
+Let's break down the key technologies we're using:
+
+* Hono: A lightweight web framework specifically designed for edge computing platforms like Cloudflare Workers
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
* [Hono](https://hono.dev/): A lightweight web framework specifically designed for edge computing platforms like Cloudflare Workers
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+* D1: Cloudflare's serverless SQL database, perfect for storing our monitoring results
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
* [D1](https://developers.cloudflare.com/d1/): Cloudflare's serverless SQL database, perfect for storing our monitoring results
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+* Drizzle ORM: A TypeScript-first ORM that provides type safety and a great developer experience
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
* [Drizzle ORM](https://orm.drizzle.team/): A TypeScript-first ORM that provides type safety and a great developer experience
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+* Durable Objects: Cloudflare's solution for maintaining state and coordinating our monitoring schedules
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
* [Durable Objects](https://developers.cloudflare.com/durable-objects/): Cloudflare's solution for maintaining state and coordinating our monitoring schedules
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx (hunk 1 above)
Comment (Nlea) [reply www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]: I can't see Otel as part of the application code? Or should this relate to fiberplane's hono client that uses Otel?

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+## Building the Monitor Class (Durables Object)
+
+The Monitor class is the core engine of our uptime monitoring system.
+What makes it special is that it's implemented as a Cloudflare Durable Object, which means it maintains its state across multiple executions.
+Each website gets its own Monitor instance that persists in memory, keeping track of its own timer and check schedule.
+The `scheduleChecks` method sets up an interval timer based on the website's configured check frequency, while `performCheck` does the actual work of making HTTP requests to the target site, measuring response times, and storing results.
+This design ensures reliable, consistent monitoring even at scale. If a website is configured to be checked every 60 seconds, the Monitor instance will maintain that schedule without requiring external coordination.
+
+```typescript
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
```typescript, titel: monitor.ts
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+We also implement an endpoint to calculate uptime percentages for any given website.
+The uptime percentage calculation is more sophisticated than it might appear at first glance.
+Rather than just storing a running total, it calculates uptime dynamically based on historical check data.
+This allows for flexible time windows - you can calculate uptime for the last day, week, or month without needing to store separate aggregates.
+The calculation uses SQL's date handling capabilities to filter checks within the specified time period, then simply divides the number of successful checks by total checks.
+What's clever about this approach is that it automatically accounts for varying check intervals - a site checked every minute will have more data points than one checked every hour, but both will give accurate uptime percentages.
+The results are rounded to two decimal places to provide meaningful precision without false accuracy.
+
+```typescript
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
```typescript, titel: index.tsx
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+  return c.json({
+    uptimePercentage: Math.round(uptimePercentage * 100) / 100,
+    period: `${days} days`
+  });
+});
+```
+
+## The Web Interface
+
+Finally, we create a simple web interface to view monitored websites and their uptime percentages.
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]: Maybe link to Oscar's blogpost here

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+The frontend of the uptime monitor is built using Hono's JSX rendering capabilities, featuring two main components: a `Layout` component and a `WebsiteList` component.
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]:
```suggestion
The frontend of the uptime monitor is built using Hono's JSX rendering capabilities, featuring two main [components](https://github.com/fiberplane/create-honc-app/tree/main/examples/uptime-monitor/src/components): a `Layout` component and a `WebsiteList` component.
```

File: www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+The `Layout` component serves as a base template, providing a clean, flexible structure with CSS-in-JS styling that ensures full viewport height and proper padding.
+The `WebsiteList` component is where the main UI lives, displaying monitored websites in a responsive grid layout
+
```
Comment (Nlea) [inline www/src/content/blog/2024-12-12-honc-uptime-monitor.mdx]: Do you have a screenshot of the UI or is this worker somewhere live and we can share a link to it

**Review by Nlea — COMMENTED** [review]
looks good to me I included some links to resources and add a few suggestions

**Review by laulauland — APPROVED** [review]

**Pull request merged**